presto-main/src/main/java/com/facebook/presto/sql/planner/DomainTranslator.java

```diff
@@ -692,34 +689,101 @@ protected ExtractionResult visitNullLiteral(NullLiteral node, Boolean complement
 
         Type leftType = expressionTypes.get(comparison.getLeft());
         Type rightType = expressionTypes.get(comparison.getRight());
-
         Preconditions.checkArgument(leftType.equals(rightType), "left and right type do not match in comparison expression (%s)", comparison);
 
-        if (left instanceof SymbolReference && !(right instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) left, comparison.getType(), new NullableValue(rightType, right)));
+        if (left instanceof Expression && right instanceof Expression) {
+            return Optional.empty();
+        }
+
+        if (!(left instanceof Expression) && !(right instanceof Expression)) {
+            return Optional.empty();
+        }
+
+        Expression symbolExpression;
+        ComparisonExpressionType comparisonType;
+        NullableValue value;
+
+        if (left instanceof Expression) {
+            symbolExpression = comparison.getLeft();
+            comparisonType = comparison.getType();
+            value = new NullableValue(rightType, right);
+        }
+        else {
+            symbolExpression = comparison.getRight();
+            comparisonType = comparison.getType().flip();
+            value = new NullableValue(leftType, left);
+        }
+
+        if (!checkSymbolExpressionValid(metadata, expressionTypes, symbolExpression)) {
+            return Optional.empty();
         }
-        if (right instanceof SymbolReference && !(left instanceof Expression)) {
-            return Optional.of(new NormalizedSimpleComparison((SymbolReference) right, comparison.getType().flip(), new NullableValue(leftType, left)));
+
+        return Optional.of(new NormalizedSimpleComparison(symbolExpression, comparisonType, value));
+    }
+
+    private static Type typeOf(Expression expression, Session session, Metadata metadata, Map<Symbol, Type> types)
+    {
+        IdentityHashMap<Expression, Type> expressionTypes = ExpressionAnalyzer.getExpressionTypes(session, metadata, new SqlParser(), types, expression, emptyList() /* parameters already replaced */);
+        return expressionTypes.get(expression);
+    }
+
+    private static boolean checkSymbolExpressionValid(Metadata metadata, IdentityHashMap<Expression, Type> expressionTypes, Expression symbolExpression)
+    {
+        while (true) {
+            if (symbolExpression instanceof SymbolReference) {
+                return true;
+            }
+            else if (symbolExpression instanceof Cast) {
+                if (!isCoercion(metadata, expressionTypes, (Cast) symbolExpression)) {
```

Comment:
Actually this case should work fine.
Tests for cases like that (though not exactly BIGINT vs DOUBLE) are in testNumericTypeTranslation.

Such expression would be rewritten to expression which is false for all non-null values in `my_bigint`.
Because `1.1 != cast(floor_cast(1.1))`.

Take a look at section:
```
case EQUAL: {
                    if (coercedValueIsEqualToOriginal) {
                        return new ComparisonExpression(EQUAL, symbolExpression, coercedLiteral);
                    }
                    // Return something that is false for all non-null values
                    return and(new ComparisonExpression(EQUAL, symbolExpression, coercedLiteral),
                            new ComparisonExpression(NOT_EQUAL, symbolExpression, coercedLiteral));
                }
```
in `rewriteComparisonExpression`.

Actually I am not sure if we should keep `NULL` values here. Or rather return empty domain for such case.
Is TupleDomain used in contexts where we want to distinguish `NULL` and `FALSE` as result from predicate?